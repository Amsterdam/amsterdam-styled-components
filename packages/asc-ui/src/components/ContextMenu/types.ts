enum Position {
  bottom = 'bottom',
  top = 'top',
}

export default {
  Position,
}
