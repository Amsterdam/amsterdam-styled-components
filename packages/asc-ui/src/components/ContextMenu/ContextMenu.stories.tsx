import React from 'react'
import { storiesOf } from '@storybook/react'
import { ReactComponent as Ellipsis } from '@datapunt/asc-assets/lib/Icons/Ellipsis.svg'
import { ReactComponent as FacebookPadded } from '@datapunt/asc-assets/lib/Icons/FacebookPadded.svg'
import { ReactComponent as Twitter } from '@datapunt/asc-assets/lib/Icons/Twitter.svg'
import { ReactComponent as Linkedin } from '@datapunt/asc-assets/lib/Icons/Linkedin.svg'
import { ReactComponent as Email } from '@datapunt/asc-assets/lib/Icons/Email.svg'
import { ReactComponent as Print } from '@datapunt/asc-assets/lib/Icons/Print.svg'
import { ReactComponent as ChevronDown } from '@datapunt/asc-assets/lib/Icons/ChevronDown.svg'
import { AscCore } from '../../styles'
import { ContextMenu, ContextMenuItem, Icon } from '../..'

storiesOf('Composed/ContextMenu', module)
  .add('default', () => (
    <ContextMenu arrowIcon={<ChevronDown />}>
      <ContextMenuItem role="button" onClick={() => {}}>
        One
      </ContextMenuItem>
      <ContextMenuItem role="button" onClick={() => {}}>
        Two
      </ContextMenuItem>
    </ContextMenu>
  ))
  .add('position bottom', () => (
    <ContextMenu
      arrowIcon={<ChevronDown />}
      position={AscCore.ContextMenuTypes.Position.bottom}
    >
      <ContextMenuItem role="button" onClick={() => {}}>
        One
      </ContextMenuItem>
      <ContextMenuItem role="button" onClick={() => {}}>
        Two
      </ContextMenuItem>
    </ContextMenu>
  ))
  .add('with label', () => (
    <ContextMenu arrowIcon={<ChevronDown />} label="I'm a label!">
      <ContextMenuItem role="button" onClick={() => {}}>
        One
      </ContextMenuItem>
      <ContextMenuItem role="button" onClick={() => {}}>
        Two
      </ContextMenuItem>
    </ContextMenu>
  ))
  .add('example of actual usage', () => (
    <ContextMenu
      arrowIcon={<ChevronDown />}
      icon={
        <Icon padding={4} inline size={24}>
          <Ellipsis />
        </Icon>
      }
    >
      <ContextMenuItem
        role="button"
        divider
        onClick={() => {
          window.print()
        }}
        icon={
          <Icon padding={4} inline size={24}>
            <Print />
          </Icon>
        }
      >
        Printen
      </ContextMenuItem>
      <ContextMenuItem
        role="button"
        onClick={() => {
          window.open(
            `https://www.facebook.com/sharer/sharer.php?u=` +
              `${window.location.href}&title=${window.document.title}`,
            '_blank',
          )
        }}
        icon={
          <Icon inline size={24}>
            <FacebookPadded />
          </Icon>
        }
      >
        Deel via Facebook
      </ContextMenuItem>
      <ContextMenuItem
        role="button"
        onClick={() =>
          window.open(
            `https://twitter.com/intent/tweet?url=${window.location.href}` +
              `&text=${window.document.title}`,
            '_blank',
          )
        }
        icon={
          <Icon inline size={24} padding={4}>
            <Twitter />
          </Icon>
        }
      >
        Deel via Twitter
      </ContextMenuItem>
      <ContextMenuItem
        role="button"
        onClick={() =>
          window.open(
            `
          https://www.linkedin.com/shareArticle?url=${window.location.href}` +
              `&mini=true&title=${window.document.title}`,
            '_blank',
          )
        }
        icon={
          <Icon inline size={24} padding={4}>
            <Linkedin />
          </Icon>
        }
      >
        Deel via Linkedin
      </ContextMenuItem>
      <ContextMenuItem
        role="button"
        onClick={() => window.open(`mailto:mail@mail.com`, '_blank')}
        icon={
          <Icon inline size={24} padding={4}>
            <Email />
          </Icon>
        }
      >
        Deel via E-mail
      </ContextMenuItem>
    </ContextMenu>
  ))
