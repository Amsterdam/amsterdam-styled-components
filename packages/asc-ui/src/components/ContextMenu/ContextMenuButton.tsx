import React from 'react'
import { ReactComponent as ChevronDown } from '@datapunt/asc-assets/lib/Icons/Chevron-Down.svg'
import { AscCore } from '../../styles'
import { Icon } from '../../index'

const ContextMenuButton = ({
  id,
  label,
  position,
  icon,
  open,
  ...otherProps
}: any) => {
  return (
    <AscCore.ContextMenu.MenuButton {...{ id }} {...otherProps}>
      {icon && icon}
      {label}
      <Icon
        inline
        size={24}
        padding={4}
        rotate={
          (position === 'bottom' && !open) || (position !== 'bottom' && open)
            ? 180
            : 0
        }
      >
        <ChevronDown />
      </Icon>
    </AscCore.ContextMenu.MenuButton>
  )
}

export default ContextMenuButton
