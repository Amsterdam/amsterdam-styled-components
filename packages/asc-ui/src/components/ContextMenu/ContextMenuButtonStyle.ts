import styled from 'styled-components'
import { themeColor, focusStyleOutline, themeSpacing } from '../../utils'

const ContextMenuButtonStyle = styled.button`
  cursor: pointer;
  display: flex;
  background-color: ${themeColor('tint', 'level1')};
  align-items: center;
  height: 32px;
  padding: ${themeSpacing(0, 2)};
  justify-content: space-between;
  z-index: 1;
  border: 1px solid;
  ${focusStyleOutline()}

  & > span:not(:last-child) {
    margin-right: ${themeSpacing(2)};
  }
`

export default ContextMenuButtonStyle
