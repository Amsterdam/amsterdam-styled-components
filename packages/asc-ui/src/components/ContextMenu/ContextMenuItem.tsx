import React from 'react'
import { AscCore } from '@datapunt/asc-core'
import { KeyboardKeys } from '../../types'

type Props = {
  focused?: boolean
  onClick?: Function
  role?: string
  icon?: React.ReactNode
  divider?: boolean
}

class ContextMenuItem extends React.Component<Props> {
  root = React.createRef<HTMLDivElement>()

  componentDidUpdate() {
    const { focused } = this.props
    const ref = this.getReference('root')
    if (ref && focused) {
      ref.focus()
    }
  }

  onClick = (e: React.KeyboardEvent | React.MouseEvent) => {
    e.preventDefault()
    const { onClick } = this.props
    if (onClick) {
      onClick(e)
    }
  }

  handleKeyPress = (e: React.KeyboardEvent) => {
    if (e.key === KeyboardKeys.Enter) {
      this.onClick(e)
    }
  }

  getReference = (el: string) => {
    if (this[el].current) {
      return this[el].current
    }

    return null
  }

  render() {
    const { children, focused, icon, ...otherProps }: any = this.props
    return (
      <AscCore.ContextMenu.MenuItem
        ref={this.root}
        focused={focused}
        onClick={this.onClick}
        onKeyDown={this.handleKeyPress}
        tabIndex={focused ? 0 : -1}
        {...otherProps}
      >
        {icon && icon}
        {children}
      </AscCore.ContextMenu.MenuItem>
    )
  }
}

export default ContextMenuItem
