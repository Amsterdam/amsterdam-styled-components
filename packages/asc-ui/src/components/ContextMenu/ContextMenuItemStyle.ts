import styled, { css } from '@datapunt/asc-core'
import { ContextMenuItemProps } from './types'
import { color, fromTheme } from '../../utils'

export const ContextMenuItemStyle = styled.li<ContextMenuItemProps>`
  padding: 0 15px 0 0;
  margin: 0;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 34px;
  width: 100%;
  cursor: pointer;
  font-size: ${fromTheme('typography.fontSize')};
  text-decoration: none;
  color: ${color('tint', 'level7')};

  & > span:first-child {
    margin: 5px 6px;
  }

  ${({ divider }) =>
    divider &&
    css`
      border-bottom: 1px solid ${color('tint', 'level4')};
    `}

  &:hover,
  &:focus {
    outline: none;
    background-color: ${color('tint', 'level2')};
  }
`

export default ContextMenuItemStyle
