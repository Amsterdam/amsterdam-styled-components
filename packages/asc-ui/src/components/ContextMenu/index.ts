export { default } from './ContextMenu'
export { default as ContextMenuSelect } from './ContextMenuSelect'
