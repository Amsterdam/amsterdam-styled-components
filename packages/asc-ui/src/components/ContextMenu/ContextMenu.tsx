import React from 'react'
import { AscCore } from '../../styles'
import ContextMenuButton from './ContextMenuButton'
import MenuList from './ContextMenuList'
import { KeyboardKeys } from '../../types'
import ownerDocument from '../../utils/ownerDocument'

const selectedChildInitial = -1

type Props = {
  position?: AscCore.ContextMenuTypes.Position
  label?: string
  icon?: React.ReactNode
  open?: boolean
}

type State = {
  open?: boolean
  selectedChild: number
}

class ContextMenu extends React.Component<Props, State> {
  static defaultProps = {
    open: false,
  }

  constructor(props: Props) {
    super(props)

    this.state = {
      open: props.open,
      selectedChild: selectedChildInitial,
    }
  }

  componentDidUpdate(prevProps: Props) {
    const { open } = this.props;

    if (prevProps.open !== open) {
      this.setState({ open })
    }
  }

  list = React.createRef<HTMLDivElement>()
  root = React.createRef<HTMLDivElement>()

  onKeyDown = (event: React.KeyboardEvent) => {
    const { children } = this.props
    const { selectedChild, open } = this.state

    const nrOfChildren = React.Children.count(children)

    if (!open) {
      return
    }

    const firstChild = 0
    const lastChild = nrOfChildren - 1

    if (event.key === KeyboardKeys.ArrowDown) {
      event.preventDefault()
      this.setState({
        selectedChild:
          selectedChild === lastChild ? firstChild : selectedChild + 1,
      })
    }

    if (event.key === KeyboardKeys.ArrowUp) {
      event.preventDefault()
      this.setState({
        selectedChild:
          selectedChild === firstChild || selectedChild === selectedChildInitial
            ? lastChild
            : selectedChild - 1,
      })
    }
  }

  onToggle = () => {
    const { open } = this.state
    this.setState({
      open: !open,
    })
  }

  onClose = () => {
    setTimeout(() => {
      const element = this.getReference('root') as HTMLInputElement
      if (element) {
        const currentFocus = ownerDocument(element).activeElement
        if (!element.contains(currentFocus)) {
          this.setState({
            selectedChild: selectedChildInitial,
            open: false,
          })
        }
      }
    })
  }

  getReference = (el: string) => {
    if (this[el].current) {
      return this[el].current
    }

    return null
  }

  render() {
    const { id, label, children, position, icon }: any = this.props
    const { open, selectedChild } = this.state

    console.log('open', open, this.props.open)

    return (
      <AscCore.ContextMenu.MenuWrapper
        id={id}
        ref={this.root}
        onKeyDown={this.onKeyDown}
        onBlur={this.onClose}
      >
        <ContextMenuButton
          {...{
            icon,
            open,
            position,
            label,
          }}
          onClick={this.onToggle}
        />
        <MenuList
          {...{
            position,
            id,
            open,
            selectedChild,
          }}
          onClose={this.onClose}
          ref={this.list}
        >
          {children}
        </MenuList>
      </AscCore.ContextMenu.MenuWrapper>
    )
  }
}

export default ContextMenu
