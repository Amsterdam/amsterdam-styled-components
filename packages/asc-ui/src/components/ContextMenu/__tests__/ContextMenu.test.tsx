import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { ChevronDown } from '@datapunt/asc-assets'
import ContextMenu from '../ContextMenu'
import ContextMenuItem from '../ContextMenuItem'
import { KeyboardKeys } from '../../../types'

jest.useFakeTimers()

describe('ContextMenu', () => {
  let component: ShallowWrapper<any, any>
  let instance: any
  let label: ShallowWrapper
  let wrapper: ShallowWrapper

  beforeEach(() => {
    component = shallow<ContextMenu>(
      <ContextMenu arrowIcon={<ChevronDown />} label="Click on me">
        <ContextMenuItem>One</ContextMenuItem>
        <ContextMenuItem>Two</ContextMenuItem>
        <ContextMenuItem>Three</ContextMenuItem>
      </ContextMenu>,
    )
    instance = component.instance()
    label = component.find('ContextMenuButton')
    wrapper = component.at(0)
  })

  it('should render', () => {
    expect(component).toMatchSnapshot()
  })

  describe('click and blur', () => {
    it('should toggle the open state', () => {
      expect(instance.state.open).toBe(false)

      label.at(0).simulate('click')
      expect(instance.state.open).toBe(true)
    })

    it('should set the open state from props', () => {
      component.setProps({ open: true })

      expect(instance.state.open).toBe(true)
    })

    it('should set the open state to false and reset the selectedChild', () => {
      component.setState({
        open: true,
      })
      jest.spyOn(instance, 'getReference').mockImplementation(() => ({
        contains: () => false,
      }))

      wrapper.simulate('blur')
      jest.runAllTimers()

      expect(instance.state.open).toBe(false)
      expect(instance.state.selectedChild).toBe(-1)
    })
  })

  describe('keyboard event', () => {
    const preventDefaultMock = {
      preventDefault: () => {},
    }
    const arrowDown = {
      ...preventDefaultMock,
      key: KeyboardKeys.ArrowDown,
    }

    const arrowUp = {
      ...preventDefaultMock,
      key: KeyboardKeys.ArrowUp,
    }
    it('should ignore any actions when state is !open', () => {
      component.setState({ open: false })
      wrapper.simulate('keydown', arrowDown)
      expect(instance.state.selectedChild).toBe(-1)
    })
    it('should increment selectedChild state when pressing ArrowDown key', () => {
      component.setState({
        open: true,
        selectedChild: 0,
      })
      wrapper.simulate('keydown', arrowDown)
      expect(instance.state.selectedChild).toBe(1)
    })

    it('should decrement selectedChild state when pressing ArrowUp key', () => {
      component.setState({
        open: true,
        selectedChild: 2,
      })
      wrapper.simulate('keydown', arrowUp)
      expect(instance.state.selectedChild).toBe(1)
    })

    it('should jump to the first menu item when pressing ArrowDown on the last item', () => {
      component.setState({
        open: true,
        selectedChild: 2,
      })
      wrapper.simulate('keydown', arrowDown)
      expect(instance.state.selectedChild).toBe(0)
    })

    it('should jump to the last menu item when pressing ArrowUp on the first or unset (-1) item', () => {
      component.setState({
        open: true,
      })
      wrapper.simulate('keydown', arrowUp)
      expect(instance.state.selectedChild).toBe(2)

      component.setState({
        selectedChild: 0,
      })
      wrapper.simulate('keydown', arrowUp)
      expect(instance.state.selectedChild).toBe(2)
    })
  })
})
