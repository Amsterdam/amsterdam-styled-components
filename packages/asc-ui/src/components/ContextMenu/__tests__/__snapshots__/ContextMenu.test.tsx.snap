// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu should render 1`] = `
<styled.div
  onBlur={[Function]}
  onKeyDown={[Function]}
>
  <ContextMenuButton
    arrowIcon={<ReactComponent />}
    label="Click on me"
    onClick={[Function]}
    open={false}
  />
  <ForwardRef
    onClose={[Function]}
    open={false}
    selectedChild={-1}
  >
    <ContextMenuItem>
      One
    </ContextMenuItem>
    <ContextMenuItem>
      Two
    </ContextMenuItem>
    <ContextMenuItem>
      Three
    </ContextMenuItem>
  </ForwardRef>
</styled.div>
`;
